### ISSUE TITLE
InstanceNorm and GraphNorm are extremely slow when using batch indices

### ISSUE THREAD

**Issue by johnpeterflynn** [pyg-team/pytorch_geometric#3503]
## 🐛 Bug

Both InstanceNorm and GraphNorm accept an optional batch index tensor in their forward passes. Unfortunately using this parameter results in a dramatic slowdown in the backward pass on lines of code that use batch as an index tensor, like [this line](https://github.com/pyg-team/pytorch_geometric/blob/cca3c0cb1fec2e60e56465ec5efb18fe5d09797d/torch_geometric/nn/norm/instance_norm.py#L65) in InstanceNorm.

The problem is that the backward pass through vector indexing launches kernels that execute serially for each index. Since the batch tensor is full of repeating indices, this results in a long list of kernels that wait for wach other to execute. It's actually a PyTorch problem that has yet to be fixed (see [#41162](https://github.com/pytorch/pytorch/issues/41162)). However this problem is exaggerated in pytorch_geometric because of the way we handle batches. 

## To Reproduce

1. Build a network to process any graph with batch size > 1.
2. Use InstanceNorm and pass to it the batch indices that are auto-generated in a Data object.
3. Benchmark performance against InstanceNorm without passing to it the batch indices.

## Expected behavior

Backward pass through tensor indexing should be the same speed no matter what the index values are.

## Environment

* OS: Ubuntu 20.04
* Python version: 3.6
* PyTorch version: 1.10
* CUDA/cuDNN version: 11.1
* GCC version:
* Any other relevant information:

## Additional context

There is a rather simple fix (which might just be a band-aid). In any location where tensor indexing is used, replace it with `torch.index_select()`.

For example, replace `x = x - mean[batch]` with `x = x - torch.index_select(mean, dim=0, batch)`.

**Comment by rusty1s** [comment]
Script to reproduce:
```python
import time

import torch
from torch.nn import Linear
import torch.nn.functional as F

from torch_geometric.loader import DataLoader
from torch_geometric.datasets import TUDataset
from torch_geometric.nn import global_mean_pool, InstanceNorm

dataset = TUDataset('/tmp/TU', name='PROTEINS')
loader = DataLoader(dataset, batch_size=len(dataset))
data = next(iter(loader)).cuda()


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.lin1 = Linear(dataset.num_features, 1024)
        self.norm = InstanceNorm(1024)
        self.lin2 = Linear(1024, dataset.num_classes)

    def forward(self, x, batch):
        x = self.lin1(x).relu()
        x = self.norm(x, batch)
        x = global_mean_pool(x, batch)
        return self.lin2(x)


model = Model().cuda()
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
criterion = torch.nn

for epoch in range(1, 101):
    torch.cuda.synchronize()
    t = time.perf_counter()
    optimizer.zero_grad()
    out = model(data.x, data.batch)
    loss = F.cross_entropy(out, data.y)
    loss.backward()
    optimizer.step()
    torch.cuda.synchronize()
    t = time.perf_counter() - t

    print(f'Epoch: {epoch:03d}, Loss: {float(loss):.4f}, Time: {t:.4f}')
```

**Comment by rusty1s** [comment]
I see that the `mean[batch]` call does slow down program execution, but it doesn't look like replacing it with `index_select` helps. Can you confirm?

**Comment by johnpeterflynn** [comment]
The graph you build is relatively small and has many batches. It's more obvious when we take batch indexing to the extreme (say just two batches, so batch index tensor is full of ones and zeros). Apologies, I should have provided some working code from the start. However I modified your program to reproduce the issue :).

On my machine this program took 0.0165 seconds per iteration for mean.index_select(0, batch) and 0.0289 seconds for mean[batch].

```python
import time

import torch
from torch.nn import Linear
import torch.nn.functional as F

from torch_geometric.loader import DataLoader
from torch_geometric.nn import global_mean_pool

num_features = 3
num_classes = 2
num_nodes = 128 * 128 
num_batches = 2
x = torch.rand((num_batches * num_nodes, num_features)).cuda()
y = torch.randint(0, num_classes, (num_batches * num_nodes,)).cuda()
batch = torch.arange(0, num_batches, dtype=torch.long).repeat_interleave(num_nodes).cuda()

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.lin1 = Linear(dataset.num_features, 1024)
        self.lin2 = Linear(1024, dataset.num_classes)

    def forward(self, x, batch):
        x = self.lin1(x).relu()
        mean = global_mean_pool(x, batch)
        if False:
            x = x - mean.index_select(0, batch)
        else:
            x = x - mean[batch]
        return self.lin2(x)


model = Model().cuda()
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
criterion = torch.nn

for epoch in range(1, 101):
    torch.cuda.synchronize()
    t = time.perf_counter()
    optimizer.zero_grad()
    out = model(x, batch)
    loss = F.cross_entropy(out, y)
    loss.backward()
    optimizer.step()
    torch.cuda.synchronize()
    t = time.perf_counter() - t
    print(f'Epoch: {epoch:03d}, Loss: {float(loss):.4f}, Time: {t:.4f}')

```

**Comment by rusty1s** [comment]
Yeah, you are right. Really interesting. As a temporary fix, I replaced all `x[batch]` calls with `x.index_select(0, batch)`, see https://github.com/pyg-team/pytorch_geometric/commit/75e04e34e840343eefd32d6be7ccd5fd4d2e560c.
